flytepropeller/pkg/controller/nodes/common/utils.go

```diff
@@ -67,3 +78,89 @@ func GetTargetEntity(ctx context.Context, nCtx interfaces.NodeExecutionContext)
+	literalSizeBytes := int64(proto.Size(toBeOffloaded))
+	literalSizeMB := literalSizeBytes / MB
+	// check if the literal is large
+	if literalSizeMB >= literalOffloadingConfig.MaxSizeInMBForOffloading {
+		errString := fmt.Sprintf("Literal size [%d] MB is larger than the max size [%d] MB for offloading", literalSizeMB, literalOffloadingConfig.MaxSizeInMBForOffloading)
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+	if literalSizeMB < literalOffloadingConfig.MinSizeInMBForOffloading {
+		logger.Debugf(ctx, "Literal size [%d] MB is smaller than the min size [%d] MB for offloading", literalSizeMB, literalOffloadingConfig.MinSizeInMBForOffloading)
+		return nil
+	}
+
+	inferredType := validators.LiteralTypeForLiteral(toBeOffloaded)
+	if inferredType == nil {
+		errString := "Failed to determine literal type for offloaded literal"
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+
+	// offload the literal
+	if err := datastore.WriteProtobuf(ctx, dataReference, storage.Options{}, toBeOffloaded); err != nil {
+		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
+		return err
+	}
+
+	// update the literal with the offloaded URI, size and inferred type
+	toBeOffloaded.Value = &idlcore.Literal_OffloadedMetadata{
+		OffloadedMetadata: &idlcore.LiteralOffloadedMetadata{
+			Uri:          dataReference.String(),
+			SizeBytes:    uint64(literalSizeBytes),
+			InferredType: inferredType,
+		},
+	}
+	logger.Infof(ctx, "Offloaded literal at location [%s] with size [%d] MB and inferred type [%s]", dataReference, literalSizeMB, inferredType)
+	return nil
+}
+
+// CheckOffloadingCompat checks if the upstream and downstream nodes are compatible with the literal offloading feature and returns an error if not contained in phase info object
+func CheckOffloadingCompat(ctx context.Context, nCtx interfaces.NodeExecutionContext, inputLiterals map[string]*core.Literal, node v1alpha1.ExecutableNode, literalOffloadingConfig config.LiteralOffloadingConfig) *handler.PhaseInfo {
+	consumesOffloadLiteral := false
+	for _, val := range inputLiterals {
+		if val != nil && val.GetOffloadedMetadata() != nil {
+			consumesOffloadLiteral = true
+			break
+		}
+	}
+	if !consumesOffloadLiteral {
+		return nil
+	}
+	var phaseInfo handler.PhaseInfo
+
+	// Return early if the node is not of type NodeKindTask
+	if node.GetKind() != v1alpha1.NodeKindTask {
+		logger.Warnf(ctx, "literal offloading : skipping sdk version check for node kind [%s'] with id [%v] ", node.GetKind(), node.GetID())
+		return nil
+	}
+
+	// Process NodeKindTask
+	taskID := *node.GetTaskID()
+	taskNode, err := nCtx.ExecutionContext().GetTask(taskID)
+	if err != nil {
+		phaseInfo = handler.PhaseInfoFailure(core.ExecutionError_SYSTEM, "GetTaskIDFailure", err.Error(), nil)
+		return &phaseInfo
+	}
+	runtimeData := taskNode.CoreTask().GetMetadata().GetRuntime()
```

Comment:
I don't have good context here, but TaskNodes encompass all task executions in Flyte (actors, agents, spark, dask, ray, etc) how is this `runtimeData` set for each? For example, if we offload into a agent task is there a specific check we need to do there?

Reply:
if my understanding is correct, anything we serialize from flytekit should have the runtime set from the SDK, right? 

e.g. here for spark: https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/plugins/flytekit-spark/flytekitplugins/spark/task.py#L153-L159 we call the super method for `PythonFunctionTask`

this https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/flytekit/core/python_function_task.py#L81 inherits from `PythonAutoContainerTask`

which itself inherits from `PythonTask` (https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/flytekit/core/python_auto_container.py#L29)

which sets the metadata https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/flytekit/core/base_task.py#L167-L169

Reply:
for agents in particular, we use the pyflyte entrypoint for all task types so I believe we should get typetransformer changes for reading offloaded inputs for free

e.g here https://github.com/flyteorg/flytekit/blob/master/flytekit/extend/backend/agent_service.py#L176 and elsewhere we call `LiteralMap.from_flyte_idl` in the agent service

we call agent.do https://github.com/flyteorg/flytekit/blob/master/flytekit/extend/backend/agent_service.py#L178

or agent.create https://github.com/flyteorg/flytekit/blob/master/flytekit/extend/backend/agent_service.py#L124

and then call `input_python_value = TypeEngine.literal_map_to_kwargs` https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/plugins/flytekit-openai/flytekitplugins/openai/chatgpt/agent.py#L33

which itself calls `to_python_value`: https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/flytekit/core/type_engine.py#L1181

which is what Eduardo's PR modifies https://github.com/flyteorg/flytekit/pull/2685/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1101

Reply:
cc @eapolinario let me know if this sounds correct to you, would love someone from OSS to verify 🙂

Reply:
cc : @wild-endeavor @pingsutw

Reply:
> anything we serialize from flytekit should have the runtime set from the SDK, right?

yes, you should be able to see these metadata on flyteconsole for every kind of task.

However, this is the Flytekit version you use locally, and it doesn't match the one you use in the image.

In other words, users may use flytekit 1.13.5 to serialize the workflow, but use flytekit 1.12 in the container.

![Screenshot 2024-09-12 at 2 45 29 PM](https://github.com/user-attachments/assets/3a294a52-1024-41dc-afb6-da6ec865d9e2)

Reply:
> for agents in particular, we use the pyflyte entrypoint for all task types so I believe we should get typetransformer changes for reading offloaded inputs for free

yes, type transformer is used to ser/de literal, so we should only change the transformer.

https://github.com/flyteorg/flytekit/blob/c06ef30518dec2057e554fbed375dfa43b985c60/plugins/flytekit-bigquery/flytekitplugins/bigquery/agent.py#L52

Reply:
thanks @pingsutw I think following the implementation of `literal_map_to_kwargs` which calls

to_python_value: https://github.com/flyteorg/flytekit/blob/ae9c6f8de21eb716dbda8d1af555c2057655db1d/flytekit/core/type_engine.py#L1181

this does trigger Eduardo's modified code path https://github.com/flyteorg/flytekit/pull/2685/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1101

 
the more concerning bit is that the runtime metadata doesn't correspond to the container image. is the only way to verify the SDK version using the code to introspect the image then?

Reply:
>users may use flytekit 1.13.5 to serialize the workflow, but use flytekit 1.12 in the container.

Which should be fine , the version checks make sure of the compatibility 


>yes, type transformer is used to ser/de literal, so we should only change the transformer.
Given the above path @katrogan pointed out, i dont see we need to change anything here as the base changes already take care of it.

Reply:
> the more concerning bit is that the runtime metadata doesn't correspond to the container image. is the only way to verify the SDK version using the code to introspect the image then?

I think so. We could probably release flytekit first and ensure most users have upgraded it.

Reply:
We handle the flytekit version check, so IMO it shouldn't block this PR and if folks are ok with the current version of the code then lets check this in unless there are any other concerns. @hamersaw @katrogan @pingsutw

Reply:
sounds good to me, I think this is the best check we have for compatibility